svm/src/account_loader.rs

```diff
@@ -95,6 +95,262 @@ pub struct FeesOnlyTransaction {
+                loaded_size: account.data().len(),
+                account,
+                rent_collected: 0,
+            })
+        } else {
+            self.callbacks
+                .inspect_account(account_key, AccountState::Dead, is_writable);
+
+            None
+        }
+    }
+
+    pub fn update_accounts_for_executed_tx(
+        &mut self,
+        message: &impl SVMMessage,
+        processed_transaction: &ProcessedTransaction,
+    ) {
+        match processed_transaction {
+            ProcessedTransaction::Executed(executed_tx) => {
+                if executed_tx.execution_details.status.is_ok() {
+                    self.update_accounts_for_successful_tx(
+                        message,
+                        &executed_tx.loaded_transaction.accounts,
+                    );
+                } else {
+                    self.update_accounts_for_failed_tx(
+                        message,
+                        &executed_tx.loaded_transaction.rollback_accounts,
+                    );
+                }
+            }
+            ProcessedTransaction::FeesOnly(fees_only_tx) => {
+                self.update_accounts_for_failed_tx(message, &fees_only_tx.rollback_accounts);
+            }
+        }
+    }
+
+    pub fn update_accounts_for_failed_tx(
+        &mut self,
+        message: &impl SVMMessage,
+        rollback_accounts: &RollbackAccounts,
+    ) {
+        let fee_payer_address = message.fee_payer();
+        match rollback_accounts {
+            RollbackAccounts::FeePayerOnly { fee_payer_account } => {
+                self.update_account(fee_payer_address, fee_payer_account);
+            }
+            RollbackAccounts::SameNonceAndFeePayer { nonce } => {
+                self.update_account(nonce.address(), nonce.account());
+            }
+            RollbackAccounts::SeparateNonceAndFeePayer {
+                nonce,
+                fee_payer_account,
+            } => {
+                self.update_account(nonce.address(), nonce.account());
+                self.update_account(fee_payer_address, fee_payer_account);
+            }
+        }
+    }
+
+    fn update_accounts_for_successful_tx(
+        &mut self,
+        message: &impl SVMMessage,
+        transaction_accounts: &[TransactionAccount],
+    ) {
+        for (i, (address, account)) in (0..message.account_keys().len()).zip(transaction_accounts) {
+            if !message.is_writable(i) {
+                continue;
+            }
+
+            // Accounts that are invoked and also not passed as an instruction
+            // account to a program don't need to be stored because it's assumed
+            // to be impossible for a committable transaction to modify an
+            // invoked account if said account isn't passed to some program.
+            if message.is_invoked(i) && !message.is_instruction_account(i) {
+                continue;
+            }
+
+            self.update_account(address, account);
+        }
+    }
```

Comment:
this and the two related functions `update_accounts_for_failed_tx` and `update_accounts_for_executed_tx` are stripped-down versions of code from `account_saver.rs`. we talked before about possibly modularizing this code so it isnt duplicated, but i couldnt find a reasonable way to do that because the account saver version expects to `move` the execution result and read/writes several vectors of whole-transaction state at its deepest levels

the trickiest thing this code did tho was the `.filter(|m| m.is_writable(i) || {!m.is_invoked(i) && m.is_instruction_account(i) })` thing, but i changed it to be more straightforward (also in monorepo account saver)